### PR TITLE
Layout propagation

### PR DESCRIPTION
**Pull request by jjsjann123**
Stacked PRs:
#1755 enabling layout propagation through runtime
#1792 propagation rule for broadcast
#1790 propagation rule for binary op
==== #1788 adding layout inference pass **_<- this one_**

What's in this PR:
inferenceAllocationOrder pass that works on an entire Fusion:
It computes AllocationOrder on inputs by looking at each TensorView's allocation_domain and rfactor_domain;
It uses a predefined rule (in AllocationOrderInferencer) to traverse and propagate AllocationOrder from inputs to the entire fusion;

Note that the pass itself doesn't mutate the fusion IR. It's just a utility function that suggests ways to specify allocation domain to be used by other optimization passes.

- [x] adding inferenceAllocationOrder pass function;
- [x] adding propagate rule for unary op;
- [x] adding cpp test to verify propagation rule;

Quick design doc: https://github.com/NVIDIA/Fuser/issues/1756

Future Work:
* expanding propagation rule to cover more operation;

### CODE REVIEW

**Comment by jjsjann123** [comment]
!build

File: csrc/preseg_passes/allocation_order_inference.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+#include <iter_visitor.h>
+#include <preseg_passes/allocation_order_inference.h>
+
+namespace nvfuser {
+
+namespace {
+
+int countNonBroadcastID(const TensorView* tv) {
+  int count = 0;
+  std::for_each(
```
Comment (wujingyue) [inline csrc/preseg_passes/allocation_order_inference.cpp:19]: How about std::count_if?

File: test/test_allocation_order_inference.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  fusion.addInput(tv0);
+  auto tv1 = relu(tv0);
+  fusion.addOutput(tv1);
+
+  std::vector<IterDomain*> tv0_nhwc = {
+      tv0->axis(0), tv0->axis(2), tv0->axis(3), tv0->axis(1)};
+  tv0->setAllocationDomain(tv0_nhwc, true);
+
+  auto updated_layout = inferenceAllocationOrder(&fusion);
+  EXPECT_THAT(updated_layout[tv1], ElementsAre(0, 2, 3, 1));
```
Comment (wujingyue) [inline test/test_allocation_order_inference.cpp:44]:
```suggestion
  const auto inferred_layout = inferenceAllocationOrder(&fusion);
  EXPECT_THAT(updated_layout.at(tv1), ElementsAre(0, 2, 3, 1));
```

File: test/test_allocation_order_inference.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+#include <preseg_passes/allocation_order_inference.h>
+
+#include <test/utils.h>
+#include <test/validator.h>
+
+#include <torch/torch.h>
+
+namespace nvfuser {
+
+using testing::_;
```
Comment (wujingyue) [inline test/test_allocation_order_inference.cpp:24]:
```suggestion
```

**Review by wujingyue — APPROVED** [review]

File: csrc/preseg_passes/allocation_order_inference.cpp (hunk 1 above)
Comment (wujingyue) [reply csrc/preseg_passes/allocation_order_inference.cpp:19]: Also, this doesn't seem to be used until #1790. Consider moving it over there.

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
Failing test isn't related. merge as-is.

**Pull request merged**